### PR TITLE
Casting fixes for gcc 6.0

### LINKED ISSUE #6198: bin_encoder.c failed to build with GCC 6.0.0
I met the following problem while building `grpc` with gcc 6.0.0
```
src/core/ext/transport/chttp2/transport/bin_encoder.c: In function ‘grpc_chttp2_base64_encode_and_huffman_compress_impl’:
src/core/ext/transport/chttp2/transport/bin_encoder.c:197:32: error: conversion to ‘uint8_t {aka unsigned char}’ from ‘int’ may alter its value [-Werror=conversion]
     enc_add2(&out, in[0] >> 2, (uint8_t)((in[0] & 0x3) << 4) | (in[1] >> 4));
                                ^
src/core/ext/transport/chttp2/transport/bin_encoder.c:198:20: error: conversion to ‘uint8_t {aka unsigned char}’ from ‘int’ may alter its value [-Werror=conversion]
     enc_add2(&out, (uint8_t)((in[1] & 0xf) << 2) | (in[2] >> 6),
                    ^
src/core/ext/transport/chttp2/transport/bin_encoder.c:213:16: error: conversion to ‘uint8_t {aka unsigned char}’ from ‘int’ may alter its value [-Werror=conversion]
                (uint8_t)((in[0] & 0x3) << 4) | (uint8_t)(in[1] >> 4));
                ^
```

### PR DESCRIPTION
**Pull request by dgquintas**
Fixes #6198 

Checked with gcc (GCC) 6.0.0 20160305 (Red Hat 6.0.0-0.15)

### CODE REVIEW

File: src/core/ext/transport/chttp2/transport/bin_encoder.c [hunk 1]
```diff
@@ -194,9 +194,13 @@ gpr_slice grpc_chttp2_base64_encode_and_huffman_compress_impl(gpr_slice input) {
 
   /* encode full triplets */
   for (i = 0; i < input_triplets; i++) {
-    enc_add2(&out, in[0] >> 2, (uint8_t)((in[0] & 0x3) << 4) | (in[1] >> 4));
-    enc_add2(&out, (uint8_t)((in[1] & 0xf) << 2) | (in[2] >> 6),
-             (uint8_t)(in[2] & 0x3f));
+    const uint8_t low_to_high = (uint8_t)((in[0] & 0x3) << 4);
+    const uint8_t high_to_low = in[1] >> 4;
```
Comment (vjpai) [inline src/core/ext/transport/chttp2/transport/bin_encoder.c]: This line doesn't need a cast? Is it because it's always guaranteed to fit?

**Comment by vjpai** [comment]
LGTM. Will wait to see what tests say.

File: src/core/ext/transport/chttp2/transport/bin_encoder.c (hunk 1 above)
Comment (dgquintas) [inline src/core/ext/transport/chttp2/transport/bin_encoder.c]: That's my intuition as well, even though I couldn't find the authoritative rule in a quick search.

**Comment by jtattermusch** [comment]
Travis has failed.

**Comment by dgquintas** [comment]
Passing now, except for "known" flake in concurrent_connectivity_test.

**Comment by vjpai** [comment]
Still LGTM.

**Pull request merged**